### DISCUSSION TITLE
New way to replace default key-mappings

### DISCUSSION THREAD

**Discussion by machakann** [machakann/vim-sandwich#134] [Announcements]
I have made a few specification changes on key mappings recently.

## TL;DR

If you would like to remap keys, write as follows.

```vim
let g:sandwich_no_default_key_mappings = 1

" add
nmap sa <Plug>(sandwich-add)
xmap sa <Plug>(sandwich-add)
omap sa <Plug>(sandwich-add)

" delete
nmap sd <Plug>(sandwich-delete)
xmap sd <Plug>(sandwich-delete)
nmap sdb <Plug>(sandwich-delete-auto)

" replace
nmap sr <Plug>(sandwich-replace)
xmap sr <Plug>(sandwich-replace)
nmap srb <Plug>(sandwich-replace-auto)

" text-objects (if you need)
omap ib <Plug>(textobj-sandwich-auto-i)
xmap ib <Plug>(textobj-sandwich-auto-i)
omap ab <Plug>(textobj-sandwich-auto-a)
xmap ab <Plug>(textobj-sandwich-auto-a)

omap is <Plug>(textobj-sandwich-query-i)
xmap is <Plug>(textobj-sandwich-query-i)
omap as <Plug>(textobj-sandwich-query-a)
xmap as <Plug>(textobj-sandwich-query-a)
```

Change `sa`, `sd`, `sr`, `ib`, `ab`, `is`, and `as` as you like.

### Old rules

Before the commit c3877254b0058a9bc6468eace9b0cc9e2956f4ce, the default key mappings has been disabled by different options. This old rule was reasonable internally but not quite convenient for users. One is forced to define three options to disable all the default mappings. So I decided to change the way one disables those key mappings.

 - `g:operator_sandwich_no_default_key_mappings` disabled
   - `sa` in normal, visual, and operator-pending mode
   - `sd` in visual mode
   - `sr` in visual mode
 - `g:textobj_sandwich_no_default_key_mappings` disabled
   - `ib` and `ab` in visual and operator-pending mode
   - `is` and `as` in visual and operator-pending mode
 - `g:sandwich_no_default_key_mappings` disabled
   - `sd` in normal mode
   - `sr` in normal mode

### New rules

With the new rule, `g:sandwich_no_default_key_mappings` disables all the default key-mappings anyway, and the other two options are kept for partial disabling.

 - `g:operator_sandwich_no_default_key_mappings` disables
   - `sa` in normal, visual, and operator-pending mode
   - `sd` in normal and visual mode
   - `sr` in normal and visual mode
 - `g:textobj_sandwich_no_default_key_mappings` disables
   - `ib` and `ab` in visual and operator-pending mode
   - `is` and `as` in visual and operator-pending mode
 - `g:sandwich_no_default_key_mappings` disables
   - all the above keymappings

## New mapping interfaces

Before the commit b5bd9f42ed77205d937ca6a33fc290591bbbe0d2, several key mappings consisted of multiple key sequences. For example, `sd` in normal mode consisted of three key sequences.

```vim
nmap sd <Plug>(operator-sandwich-delete)<Plug>(operator-sandwich-release-count)<Plug>(textobj-sandwich-query-a)
```

However, this interface is not so convenient to remap. Now the easier interfaces listed below are available.

 - `<Plug>(sandwich-add)`
   - normal mode
   - visual mode
   - operator-pending mode
 - `<Plug>(sandwich-delete)`
   - normal mode
   - visual mode
 - `<Plug>(sandwich-replace)`
   - normal mode
   - visual mode

`sd` is then written in shorter form.

```vim
nmap sd <Plug>(sandwich-delete)
```
